Cargill/splinter
# Rework biome oauth store

Reworks the Biome OAuth user session store to match the latest designs. This includes the following changes:
* Updating the store API
* Updating the memory and DB implementations of the store
* Updating the REST API to use the new API correctly
* Replacing the `OAuthUserInfoStore` trait with direct use of the Biome `OAuthUserSessionStore`
* Adding a new `OAuthIdentityProvider` that uses the session store to get the users identity (the implementation is incomplete but required to keep things working)

### Testing

1. Run splinterd backed by postgres running in a docker container

```
$ docker run --name postgres -p 5432:5432 -e POSTGRES_PASSWORD=mypassword -e POSTGRES_DB=splinter -d postgres
$ cargo run --manifest-path cli/Cargo.toml -- database migrate -C postgres://postgres:mypassword@localhost:5432/splinter
$ mkdir -p /tmp/splinter # a splinter home dir for local use
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml -- cert generate --skip
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path splinterd/Cargo.toml --features auth,biome-oauth -- \
    --database postgres://postgres:mypassword@localhost:5432/splinter \
    --enable-biome \
    --tls-insecure \
    --oauth-provider github \
    --oauth-client-id <your client id> \
    --oauth-client-secret <your client secret> \
    --oauth-redirect-url http://localhost:8080/oauth/callback \
    -vv
```

2. In a browser, go to localhost:8080/oauth/login?redirect_url=redirect. This will result in a 401 which doesn't matter, but the browser's URL will contain the Splinter access token and your GitHub username.

3. Go to the link in step 2 again to generate a 2nd session. This will get a different access token.

4. In a new terminal window, onnect to the postgres container, login, and check the tables. There should be one user entry with two session entries for that user.

```
$ docker run -it --rm --link postgres:postgres postgres psql -h postgres -U postgres
Password for user postgres:
psql (13.0 (Debian 13.0-1.pgdg100+1))
Type "help" for help.

postgres=# \c splinter
You are now connected to database "splinter" as user "postgres".
splinter=# select * from oauth_users;
 subject  |               user_id
----------+--------------------------------------
 ltseeley | 14339126-c833-5abb-ae9a-bc9a2477f323
(1 row)

splinter=# select * from oauth_user_sessions;
      splinter_access_token       | subject  |            oauth_access_token            |
 oauth_refresh_token | last_authenticated
----------------------------------+----------+------------------------------------------+
---------------------+--------------------
 vlhBtnLl0YC5YwpM4Iec4Yy6cOar9FhZ | ltseeley | 9fe49a53de243a46478ad62dac3a88e15abab860 |
                     |         1608315746
 t9M08cTOKtWDytq8o4WW5I7XRur6Kqlp | ltseeley | 4607ec2b1c0f999625e1d3d2d00854aa2f41a00b |
                     |         1608315750
(2 rows)
```

5. In a new terminal window, verify that both Splinter access tokens can be used:

```
$ curl --header "Authorization: Bearer OAuth2:$SPLINTER_ACCESS_TOKEN1" http://localhost:8080/biome/keys
{"data":[]}
$ curl --header "Authorization: Bearer OAuth2:$SPLINTER_ACCESS_TOKEN2" http://localhost:8080/biome/keys
{"data":[]}
```
